File: src/main/java/com/example/cms/security/SecurityService.java
package com.example.cms.security;

import com.example.cms.page.Page;
import com.example.cms.university.University;
import com.example.cms.user.User;
import com.example.cms.validation.exceptions.UnauthorizedException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.session.SessionInformation;
import org.springframework.security.core.session.SessionRegistry;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@Slf4j
@RequiredArgsConstructor
public class SecurityService {
    private final SessionRegistry sessionRegistry;

    public Optional<LoggedUser> getPrincipal() {
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();

        if (principal instanceof LoggedUser) {
            return Optional.of((LoggedUser) principal);
        } else {
            return Optional.empty();
        }
    }

    public void invalidateUserSession(Long id) {
        List<Object> loggedUsers = sessionRegistry.getAllPrincipals();
        for (Object principal : loggedUsers) {
            if (principal instanceof LoggedUser) {
                final LoggedUser loggedUser = (LoggedUser) principal;
                if (loggedUser.getId().equals(id)) {
                    List<SessionInformation> sessionsInfo = sessionRegistry.getAllSessions(principal, false);
                    if (sessionsInfo != null) {
                        for (SessionInformation sessionInformation : sessionsInfo) {
                            sessionInformation.expireNow();
                            log.info(String.format("User %d: session invalidated", id));
                        }
                    }
                }
            }
        }
    }

    public boolean isForbiddenPage(Page page) {
        return getPrincipal().map(loggedUser -> {
            switch (loggedUser.getAccountType()) {
                case ADMIN:
                    return false;
                case MODERATOR:
                    return !hasUniversity(page.getUniversity().getId());
                case USER:
                    return !page.getCreator().getId().equals(loggedUser.getId()) ||
                            !hasUniversity(page.getUniversity().getId());
            }
            return true;
        }).orElse(true);
    }

    public boolean isForbiddenUniversity(University university) {
        return getPrincipal().map(loggedUser -> {
            switch (loggedUser.getAccountType()) {
                case ADMIN:
                    return false;
                case MODERATOR:
                    return !hasUniversity(university.getId());
                case USER:
                    return true;
            }
            return true;
        }).orElse(true);
    }

    public boolean isForbiddenUser(User user, boolean onlyDifferentUser) {
        return getPrincipal().map(loggedUser -> {
            if (onlyDifferentUser && loggedUser.getId().equals(user.getId())) {
                return true;
            } else {
                return isForbiddenUser(user);
            }
        }).orElse(true);
    }

    public boolean isForbiddenUser(User user) {
        return getPrincipal().map(loggedUser -> {
            switch (loggedUser.getAccountType()) {
                case ADMIN:
                    return false;
                case MODERATOR:
                    return !loggedUser.getId().equals(user.getId()) &&
                            (!hasHigherRoleThan(user.getAccountType()) ||
                                    !hasUniversity(user.getEnrolledUniversities().stream()
                                            .map(University::getId)
                                            .collect(Collectors.toList())));
                case USER:
                    return !loggedUser.getId().equals(user.getId());
            }
            return true;
        }).orElse(true);
    }

    public boolean hasUniversity(List<Long> universities) {
        LoggedUser loggedUser = getPrincipal().orElseThrow(UnauthorizedException::new);

        boolean sameUniversity = false;
        for (long universityId : universities) {
            if (loggedUser.getUniversities().contains(universityId)) {
                sameUniversity = true;
                break;
            }
        }
        return loggedUser.getAccountType().equals(Role.ADMIN) || sameUniversity;
    }

    public boolean hasUniversity(Long universityId) {
        LoggedUser loggedUser = getPrincipal().orElseThrow(UnauthorizedException::new);
        return loggedUser.getAccountType().equals(Role.ADMIN) || loggedUser.getUniversities().contains(universityId);
    }

    public boolean hasHigherRoleThan(Role userRole, Role role) {
        switch (role) {
            case ADMIN:
            case MODERATOR:
                return userRole.equals(Role.ADMIN);
            case USER:
                return userRole.equals(Role.ADMIN) || userRole.equals(Role.MODERATOR);
        }
        return false;
    }

    public boolean hasHigherRoleThan(Role role) {
        LoggedUser loggedUser = getPrincipal().orElseThrow(UnauthorizedException::new);
        return hasHigherRoleThan(loggedUser.getAccountType(), role);
    }
}

File: src/main/resources/Angular/src/assets/service/security.service.ts
import {Injectable} from '@angular/core';
import {UserService} from "./user.service";
import {Page} from "../models/page";
import {User} from "../models/user";
import {University} from "../models/university";

@Injectable({
  providedIn: 'root'
})
export class SecurityService {

  constructor(private userService: UserService) {
  }

  public isForbiddenPage(page?: Page): boolean {
    let loggedUser: User | null = this.userService.loggedUser;

    if (loggedUser && page) {
      switch (loggedUser.accountType) {
        case "ADMIN":
          return false
        case "MODERATOR":
          return !this.hasUniversity(page.university.id);
        case "USER":
          return page.creator.id !== loggedUser.id && !this.hasUniversity(page.university.id);
      }
    }

    return true;
  }

  public isForbiddenUniversity(university?: University): boolean {
    let loggedUser: User | null = this.userService.loggedUser;

    if (loggedUser && university) {
      switch (loggedUser.accountType) {
        case "ADMIN":
          return false
        case "MODERATOR":
          return !this.hasUniversity(university.id);
        case "USER":
          return true;
      }
    }

    return true;
  }


  public isForbiddenUser(user?: User, onlyDifferentUser: boolean = false): boolean {
    let loggedUser: User | null = this.userService.loggedUser;

    if (loggedUser && user) {
      if (onlyDifferentUser && loggedUser.id === user.id) {
        return true;
      }

      switch (loggedUser.accountType) {
        case "ADMIN":
          return false
        case "MODERATOR":
          return loggedUser.id !== user.id &&
            (!this.hasLoggedUserHigherRole(user.accountType) ||
            !this.hasUniversityFromList(user.enrolledUniversities.map((university) => university.id)));
        case "USER":
          return loggedUser.id !== user.id;
      }
    }


    return true;
  }


  public hasRole(role?: string): boolean {
    let loggedUser: User | null = this.userService.loggedUser;
    if (loggedUser && role) {
      switch (role) {
        case "ADMIN":
          return loggedUser.accountType === "ADMIN";
        case "MODERATOR":
          return loggedUser.accountType === "MODERATOR" || loggedUser.accountType === "ADMIN";
        case "USER":
          return true;
      }
    }
    return false;
  }

  public hasUniversityFromList(universities: number[]): boolean {
    let loggedUser: User | null = this.userService.loggedUser;

    if (loggedUser) {
      let sameUniversity: boolean = false;
      for (const university of universities) {
        if (loggedUser.enrolledUniversities.map((university) => university.id).includes(university)) {
          sameUniversity = true;
          break;
        }
      }
      return loggedUser.accountType === "ADMIN" || sameUniversity;
    }
    return false;
  }


  public hasUniversity(universityId?: number): boolean {
    let loggedUser: User | null = this.userService.loggedUser;
    if (loggedUser && universityId) {
      return loggedUser.accountType === "ADMIN" ||
        loggedUser.enrolledUniversities.map((university) => university.id).includes(universityId)
    }
    return false;
  }


  public hasUserHigherRole(userRole: string, role: string): boolean {
    switch (role) {
      case "ADMIN":
      case "MODERATOR":
        return userRole === "ADMIN";
      case "USER":
        return userRole === "ADMIN" || userRole === "MODERATOR";
    }
    return false;
  }

  public hasLoggedUserHigherRole(role: string): boolean {
    let loggedUser: User | null = this.userService.loggedUser;
    if (loggedUser) {
      return this.hasUserHigherRole(loggedUser.accountType, role)
    }
    return false;
  }
}
